chat-server/tests/e2e/viz_utils/per_example_workflow.py

```
@@ -0,0 +1,678 @@
+
+            unique_project_ids = {s.get("project_id", "") for s in schemas if s.get("project_id")}
+            unique_dataset_ids = [s.get("dataset_id", "") for s in schemas if s.get("dataset_id")]
+            unique_dataset_ids = [d for d in unique_dataset_ids if d]
+
+            test_case.project_id = next(iter(unique_project_ids)) if unique_project_ids else ""
+            test_case.dataset_id = unique_dataset_ids[0] if len(unique_dataset_ids) == 1 else ""
+
+            dataset_names_for_sql = [
+                s.get("dataset_name", "") for s in schemas if s.get("dataset_name")
+            ]
+            test_case.sql_queries = [f'SELECT * FROM "{dn}"' for dn in dataset_names_for_sql]
```

Ruff (0.14.13)

Possible SQL injection vector through string-based query construction

(S608)